### DISCUSSION TITLE
How can I correctly implement custom node equality?

### DISCUSSION THREAD

**Discussion by AndreaPra** [networkx/networkx#6742] [Q&A] [answered]
Hi everyone,

it's my first time using this library, and I am still on the moment I am doing experiments. 
I was trying to build a graph of nodes based on a custom Python `class` of mine and, since _I have a lot of nodes which I want to consider equals_. I tried looking on Internet and I ended up on an old question on Stackoverflow.

[Here](https://stackoverflow.com/questions/34731984/set-custom-node-equality-comparison-in-networkx) the Stackoverflow question I ended up to read.

### The code I tried
I tried the solution proposed in the question, I share here with you the code.
```Python
import networkx as nx
import scipy

class DictToObject:
    def __init__(self, dic):
        for key, value in dic.items():
            setattr(self, key, value)
    def __eq__(self, other):
        return self.id == other.id
    
    def __hash__(self):
        return hash(self.__dict__.values())
    
    def __repr__(self):
        return str(self.__dict__.values())
    
s1 = DictToObject({"name":"sarah", "id":"s"})
s2 = DictToObject({"name":"sarah", "id":"s"})

M = nx.Graph()
M.add_edge(s1, s2)
print(M.nodes())
print(M.edges())
print(M) # this lines throw an error on networkx/classes/reportviews.py:531, in DegreeView.__iter__(self)
```

I noticed that the graph is created without throwing any error, and it's possible to print the `edges` and the `nodes`, which are  well recognised as equals, **_but_**, I cannot do `print(M)` or `nx.draw(M)`.

 ### The error
The type of error is this one:
```python
KeyError: dict_values(['sarah', 's'])
```

If I remove the functions: `__hash__`, `__eq__` and `__repr__` it works, but of course, without the feature of considering the nodes as a set. 

I am suspicious about the fact it's the `__repr__` class that breaks everything, because the `KeyError: dict_values(['sarah', 's'])` is the output of that class. I don't know, I hope you tell me :( .

### So, the questions are
How should I implement these three functions of my node's class in order to avoid duplicate nodes?

Is there another method to achieve the same result, such as the use of the node attributes?

**Comment by dschult** [comment]
Hmmm...   What exactly do you mean by custom node equality?  If two nodes are equal, how would the dict-storage work? Wouldn't it find only 1 node instead of 2?  Said another way, if you have two nodes, they should be distinguishable. We typically distinguish nodes using equality.  It sounds like you want the nodes to be different (in our sense of the word different), but somehow have the same traits. Could you store the nodes using some identifier (maybe just an integer) and then store information about the nodes either as a node attribute, or in an external data structure with integer-node lookup capability?

**Reply by AndreaPra** [reply]
Hi, 

_thanks_ for answering me :)

you are right, I should specify better what I mean by **custom node equality.** 

- let's create my custom _class_: `DictToObject`, with implemented the function **_equality_** as I like (the `__eq__` method).
- let's create now two _objects_ out of the class we just created: `obj1 = {"name":"sarah", "id":"s"}` and `obj2 = {"name":"sarah", "id":"s"}`. 
- let's create the _nodes_ out of these two objects by using directly the `M.add_edge(s1, s2)` method.
- According to my implementation of the class they are instances of: `obj1 == obj2` returns `True`. Therefore, I am want the two nodes created are considered equals by the network graph. That's what I mean about **custom equality**.

That's why I took this code from the [stack overflow](https://stackoverflow.com/questions/34731984/set-custom-node-equality-comparison-in-networkx) question.

```python
obj1 = DictToObject({"name":"sarah", "id":"s"})
obj22 = DictToObject({"name":"sarah", "id":"s"})

M = nx.Graph()
M.add_edge(obj1, obj2)
```

To sum up, the code I have shared with you does what I want it to do: I am creating two nodes that have the same dictionary, and so, the network is considering just one node.

Indeed, the output of both `print(M.nodes())` and `print(M.edges())` is 1, as I wanted. 

The only problem is that, apparently, I cannot anymore use the `print(M)` function or draw the graph with `nx.draw(M)`; and I don't understand if it's the behaviour the creators meant or not.

I hope I make the question more clear now.

**Comment by dschult** [comment]
Got it -- You can test whether it works using a dict:
```python
s1 = DictToObject({"name":"sarah", "id":"s"})
d = {s1: 1}

s1 == next(iter(d))  # gives True.   Yay
hash(s1 == next(iter(d))  # gives True       Yay

s1 in d        # gives False           Huh?
d[s1]          # raises a KeyError because s1 is not in the dict.
```
So, if you can figure out how to make your objects work as keys in dicts, they should work for NetworkX nodes. That might give you more answers to your google searches. I thought dicts worked using equals and hash, but maybe there is something more subtle going on.

**Reply by AndreaPra** [reply]
I am sorry but I don't understand your answer.

You suggest to use the dictionaries? Because, if so, I tried and they are not working.

### First code not working
Since the python's `dict` is unhashable it cannot be add as a node to the network graph-

```python
import networkx as nx

# let's do the same thing but using dict
d1 = {"name":"sarah", "id":"s"}
d2 ={"name":"sarah", "id":"s"}

M = nx.Graph()
M.add_edge(d1, d2) # this lines throws -> TypeError: unhashable type: 'dict' 
print(M.nodes())
print(M.edges())
print(M)
```

### Second code, working, but not as I would like
By transforming the `dict` in an `iterator` it's possible to add the node to the network, still, it's not implemented the equality, meaning the network is treating the two dictionaries as two different node, which is not the behaviour I would like it to have.
```python
import networkx as nx

# let's do the same thing but using dict
d1 = iter({"name":"sarah", "id":"s"})
d2 = iter({"name":"sarah", "id":"s"})

M = nx.Graph()
M.add_edge(d1, d2)
print(M.nodes())
print(M.edges())
print(M) # Graph with 2 nodes and 1 edges
```
---
I am wondering if there is something missing in the _**documentation**_, since I tried to achieve this "custom equality" that I talked about by implementing the equality function on my `class`, but it wasn't enough. Indeed, by doing as the documentation says, I have the possibility to implement this custom equality, but I lose the possibility to **visualise** the graph. 

_**Is it possible that something is missing on the documentation in order to explain something more I should do to not to lose the possibility to visualise the graph?**_

I share more details down here about the stack trace that is printed when I run the code I showed at the beginning of the discussion:
- networkx/classes/graph.py:1957, in `Graph.number_of_edges(self, u, v)`
- networkx/classes/graph.py:1903, in `Graph.size(self, weight)`
- networkx/classes/graph.py:1903, in \<genexpr\>(.0)
- networkx/classes/reportviews.py:531, in `DegreeView.__iter__(self)`

**Reply by MridulS** [reply, marked as answer]
You still need to create the custom class `DictToObject`. The following code should work.
``` python
import networkx as nx

class DictToObject:
    def __init__(self, data):
        self._data = data

    def __eq__(self, other):
        return self._data['id'] == other._data['id']

    def __hash__(self):
        return hash(self._data['id'])

s1 = DictToObject({"name":"sarah", "id":"s"})
s2 = DictToObject({"name":"sarah", "id":"s"})

G = nx.Graph()
G.add_edge(s1, s2)

print(G)

nx.draw(G)
```

**Reply by AndreaPra** [reply]
It works, even if I don't get why. I would suggest to point this out in the documentation.

Thank you very much :)
